166 HQDM Extensions Mechanism

twalmsley opened the pull request:
See [HQDM Model Extension Example](https://github.com/twalmsley/hqdm_model_extension_example) for how to write an extension module.

(We could perhaps copy the example into the MC repo or create a better example.)

GCHQDeveloper42 commented:
This functionality could do with unit testing, but seen as we don't have a test suite for `HqdmObjectFactory` yet, that may be asking too much of this PR.

twalmsley commented:
- I added a module containing the example extension (it's a separate module so that it gets built to a JAR for use by the `examples` module).
- I wasn't able to get a JUnit test working in the `examples` module so I added it to `src/main/java` instead as a `main` program which runs successfully. There are instructions about how to run it in the class javadoc.

Perhaps at some point we can create a proper unit test if that's possible, but right now I think this might have to suffice.
